semaphoreci:
```
../_generated/linux/release/32/host_dmd -of../generated/build -g build.d
make: *** [../generated/build] Segmentation fault
```
???

--- a/src/dmd/backend/cgelem.d
+++ b/src/dmd/backend/cgelem.d
@@ -2072,6 +2072,8 @@ private elem * elcom(elem *e, goal_t goal)
 @trusted
 private elem * elcond(elem *e, goal_t goal)
 {
+    //printf("elcond() goal = %d\n", goal);
+    //elem_print(e);
     elem *e1 = e.EV.E1;
     switch (e1.Eoper)
     {
@@ -2320,10 +2322,58 @@ private elem * elcond(elem *e, goal_t goal)
                 e.EV.E1.EV.E1 = e1;
             }
 
+            /* Replace:
+             *   *p op e ? p : false
+             * with:
+             *   bool
+             */
+            else if (goal == GOALflags &&
+                ec2.Eoper == OPconst && !boolres(ec2) &&
+                typtr(ec1.Ety) &&
+                ec1.Eoper == OPvar &&
+                OTbinary(e1.Eoper) &&
+                !OTsideff(e1.Eoper) &&
+                e1.EV.E1.Eoper == OPind &&
+                el_match(findPointer(e1.EV.E1.EV.E1), ec1) &&
+                !el_sideeffect(e))
+            {
+                /* NOTE: should optimize other cases of this
+                 */
+                el_free(e.EV.E2);
+                e.EV.E2 = null;
+                e.Eoper = OPbool;
+                e.Ety = TYint;
+            }
             break;
         }
     }
     return e;
+}
+
+/******************************
+ * Given an elem that is the operand to OPind,
+ * find the expression representing the pointer.
+ * Params:
+ *      e = operand to OPind
+ * Returns:
+ *      expression that represents the pointer
+ */
+@trusted
+private elem* findPointer(elem* e)
+{
+    if (e.Eoper == OPvar)
+        return e;
+    if (OTleaf(e.Eoper) || !(e.Eoper == OPadd || e.Eoper == OPmin))
+        return null;
+
+    if (typtr(e.EV.E1.Ety))
+        return findPointer(e.EV.E1);
+    if (OTbinary(e.Eoper))
+    {
+        if (typtr(e.EV.E2.Ety))
+            return findPointer(e.EV.E2);
+    }
+    return null;
 }
 
 

--- a/test/runnable/fix22115.d
+++ b/test/runnable/fix22115.d
@@ -1,0 +1,70 @@
+/* PERMUTE_ARGS: -O -inline
+ */
+// https://issues.dlang.org/show_bug.cgi?id=22115
+
+
+int sx;
+void sss() {  ++sx; }
+
+static if (1)
+{
+    struct S { int a; }
+
+    void test1(S* s)
+    {
+        if (s.a == 3 ? s : null)
+            sss();
+    }
+}
+
+static if (1)
+{
+    extern (C++) class Exp
+    {
+        int a;
+
+        void func() { }
+        final inout(AddExp) isAddExp() inout { return a == 3 ? cast(typeof(return))this : null; }
+    }
+
+    extern (C++) class AddExp : Exp
+    {
+    }
+
+    void test2(Exp e)
+    {
+        if (e.isAddExp())
+            sss();
+    }
+}
+
+
+int main()
+{
+    static if (1)
+    {
+        S s;
+        s.a = 3;
+        test1(&s);
+        assert(sx == 1);
+        s.a = 2;
+        test1(&s);
+        assert(sx == 1);
+    }
+    sx = 1;
+
+    static if (1)
+    {
+        auto c = new AddExp();
+        c.a = 3;
+        test2(c);
+        assert(sx == 2);
+        auto ae = c.isAddExp();
+        assert(ae && ae.a == 3);
+        c.a = 2;
+        test2(c);
+        assert(sx == 2);
+    }
+
+    return 0;
+}

